spring-cloud-gcp-autoconfigure/src/main/java/com/google/cloud/spring/autoconfigure/bigquery/GcpBigQueryAutoConfiguration.java

```diff
@@ -97,25 +106,35 @@ private int getThreadPoolSize(int threadPoolSize) {
   @Bean
   @ConditionalOnMissingBean
   public BigQuery bigQuery() throws IOException {
-    BigQueryOptions bigQueryOptions =
+    BigQueryOptions.Builder bigQueryOptionsBuilder =
         BigQueryOptions.newBuilder()
             .setProjectId(this.projectId)
             .setCredentials(this.credentialsProvider.getCredentials())
-            .setHeaderProvider(new UserAgentHeaderProvider(GcpBigQueryAutoConfiguration.class))
-            .build();
-    return bigQueryOptions.getService();
+            .setHeaderProvider(new UserAgentHeaderProvider(GcpBigQueryAutoConfiguration.class));
+    if (this.universeDomain != null) {
+      bigQueryOptionsBuilder.setUniverseDomain(this.universeDomain);
+    }
+    if (this.host != null) {
+      bigQueryOptionsBuilder.setHost(this.host);
+    }
+    return bigQueryOptionsBuilder.build().getService();
   }
 
   @Bean
   @ConditionalOnMissingBean
   public BigQueryWriteClient bigQueryWriteClient() throws IOException {
-    BigQueryWriteSettings bigQueryWriteSettings =
+    BigQueryWriteSettings.Builder bigQueryWriteSettingsBuilder =
         BigQueryWriteSettings.newBuilder()
             .setCredentialsProvider(this.credentialsProvider)
             .setQuotaProjectId(this.projectId)
-            .setHeaderProvider(new UserAgentHeaderProvider(GcpBigQueryAutoConfiguration.class))
-            .build();
-    return BigQueryWriteClient.create(bigQueryWriteSettings);
+            .setHeaderProvider(new UserAgentHeaderProvider(GcpBigQueryAutoConfiguration.class));
+    if (this.universeDomain != null) {
+      bigQueryWriteSettingsBuilder.setUniverseDomain(this.universeDomain);
+    }
+    if (this.jsonWriterEndpoint != null) {
+      bigQueryWriteSettingsBuilder.setEndpoint(this.jsonWriterEndpoint);
```

Comment:
Looks like `jsonWriterEndpoint` is only used for `BigQueryWriteClient`, which is actually part of bigquery-storage, not bigquery. Ideally I think we should have different auto-configuration for them, but since they are in one auto-configuration, can we either expose `host` or `endpoint`? And construct the other dynamically?

Reply:
That's right, it's only used for BigqueryWriteClient (bigquery storage). We can expose `endpoint` and strip the port and reformat it to the pattern that [`DEFAULT_HOST`](https://github.com/googleapis/sdk-platform-java/blob/1a988df22f7e3d15ce6b121bf26897c59ab468e4/java-core/google-cloud-core/src/main/java/com/google/cloud/ServiceOptions.java#L86) currently has when setting it to BigqueryOptions. 

However, the concern I have with this is that it would prevent us from making endpoint/host configurable on a per-service basis. In fact, looking at the code again, I'm wondering if it was a mistake to make universe domain shared between bigquery and bigquerystorage.

Reply:
Do we share anything else between Bigquery and Bigquerystorage client? Ideally they should be separate as they are two different services, but looks like we are treating them the same in spring-cloud-gcp.

Reply:
That's a good question - you're right, they do seem pretty interconnected in spring-cloud-gcp.
 `BigqueryTemplate` [wraps both Bigquery and BigQueryWriteClient](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blob/f7fc095a55966a910693718deaa74d9bd2a0d2f8/spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java#L102) and both services share the `spring.cloud.gcp.bigquery.threadPoolSize` property and `spring.cloud.gcp.bigquery.datasetName` property. Dataset name is used by the [`BigtableTemplate#writeJsonStream`](https://github.com/GoogleCloudPlatform/spring-cloud-gcp/blame/f7fc095a55966a910693718deaa74d9bd2a0d2f8/spring-cloud-gcp-bigquery/src/main/java/com/google/cloud/spring/bigquery/core/BigQueryTemplate.java#L218C46-L218C55) method which uses Bigquery to create a table before invoking the BigQuery Storage Write API. 

`spring.cloud.gcp.bigquery.jsonWriterBatchSize` is the only property that is unique to bigquery storage.

Reply:
I think we have two options
1. Continue treating bigquery and bigquerystorage the same product. Then we only introduce `universe-domain`, and either `host` or `endpoint`.
2. Start treating them two different product. Then we would introduce `universe-domain` and `host` for bigquery, `universe-domain` and `endpoint` for bigquerystorage.
Any preferences? cc: @burkedavison

Reply:
As discussed offline, considering that both bigquery and bigquerystorage are wrapped as one product in spring-cloud-gcp, we will expose global endpoint and universe domain properties. The port will be omitted from the endpoint when set to the bigquery client.